### DISCUSSION TITLE
issue: `setValue` with object array makes lost of reference

### DISCUSSION THREAD

**Discussion by donrsh** [react-hook-form/react-hook-form#12044] [Q&A]
### Version Number

7.52.0

### Codesandbox/Expo snack

https://codesandbox.io/p/sandbox/react-hook-form-setvalue-with-object-array-lost-reference-vy45yy?file=%2Fpackage.json

### Steps to reproduce

Toggle an item in `MUI Select with RHF`. Observer that the item is pushed to field value every time selected.

I guess it's because when the internal `setValue` is called with array of object, something additional like cloning happened as well. Thus, the value reference is lost, which makes the MUI Select cannot decide what options are selected.

### Expected behaviour

Toggle an item in `MUI Select without RHF`. That's the expected behavior.

https://github.com/react-hook-form/react-hook-form/assets/7490144/5d040eaa-8351-4001-9f2b-ca76bd6dafa6

### What browsers are you seeing the problem on?

Chrome

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by tatsmaki** [comment]
You can do this
```
<MenuItem key={x.name} value={x.name}>
```
![Screenshot 2024-06-25 at 09 45 23](https://github.com/react-hook-form/react-hook-form/assets/61022192/4b1154d3-cad4-4f26-8686-912f85029028)

Or look for custom equality comparator in MUI Select component. I don't remember any, but here is the closed issue https://github.com/mui/material-ui/issues/24201

**Reply by donrsh** [reply]
It seems MUI team doesn't really implement that prop on `Select`. They just left it to be upvoted to decide whether to implement it (whereas closed it, that's weird ...).

Anyway, the equlity comparator you mentioned is actually implemented in `AutoComplete` ([`isOptionEqualToValue`](https://mui.com/material-ui/api/autocomplete/#autocomplete-prop-isOptionEqualToValue)). And sure, you can use it to mimic `Select`. I did implement one in the [issue codesandbox]()

Still want to know though, is there any reason for some operations that make lost of the object reference?
